Install tools from tool_list individually

mvdbeek opened the pull request:
Addresses Issue #2 and https://github.com/galaxyproject/galaxy/issues/1115
This also means that downstream project need to update their use of this role.
`galaxy_tools_tool_list` replaces `galaxy_tools_tool_list_files`.
I'd say that in case multiple tool list files are needed one could run the role for each file?!
Also [this](https://docs.ansible.com/ansible/playbooks_loops.html#loops-and-includes) is coming soon ... .

mvdbeek committed:
Install tools from tool_list individually

mvdbeek commented:
Marking WIP, because the role will still report everything to be changed. A `changed_when` should work ...

natefoo commented:
Maybe not the right time to do this, but if #2 were done, we could have better control over the changed state. I'd be happy to work on it though, if people thought it was a good idea. I've done a number of modules before and they're not terribly difficult.

afgane commented:
I was at about the same state with the code but was hoping to take your @mvdbeek comment about [nested_loops](https://docs.ansible.com/ansible/playbooks_loops.html#nested-loops) to see if multiple files can be made to work. Multiple files are a neat feature as layered tool suites can be built in a modular fashion. I'll merge this as it's the common base and keep working on the multi-file support. 

@natefoo - I thing a module would be pretty neat because running this script as an long-running, external script feels a bit off. I'd be happy to help on that effort.

mvdbeek commented:
@natefoo @afgane 
I think we can discover the state with a few tweeks to install_tool_shed_tools.py and the `changed_on`/`failed_when`.
In the current state it looks like this when mixing previsouly installed and new tools:

```
TASK: [galaxyprojectdotorg.galaxy-tools | Install Tool Shed tools] ************ 
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_picard_index_builder'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_build_kraken_database'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_fetch_ncbi_taxonomy'})
ok: [192.54.201.79] => (item={'owner': 'iuc', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_gemini_database_downloader'})
ok: [192.54.201.79] => (item={'owner': 'iuc', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_hisat2_index_builder'})
ok: [192.54.201.79] => (item={'owner': 'iuc', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_snpeff'})
ok: [192.54.201.79] => (item={'owner': 'bgruening', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_diamond_database_builder'})
ok: [192.54.201.79] => (item={'owner': 'jeremyjliu', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'region_motif_data_manager'})
changed: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_fetch_genome_all_fasta'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_rsync_g2'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_fetch_genome_dbkeys_all_fasta'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_bwa_mem_index_builder'})
changed: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_bwa_index_builder'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_bowtie2_index_builder'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_gatk_picard_index_builder'})
ok: [192.54.201.79] => (item={'owner': 'devteam', 'tool_panel_section_label': 'Data Managers for NGS Mapping', 'name': 'data_manager_sam_fasta_index_builder'})
```

It should also detect failed tools ( didn't test, someone knows one?)
For the multiple files we can re-enable those when ansible-2.0 comes out (and it's OS X support gets fixed!) .

mvdbeek committed:
Attempt to discover tool install state

afgane commented:
BTW, testing for failed tool installs can be done by supplying an invalid tool revision (and things work as expected).
Thanks.

mvdbeek commented:
Awesome, good to know, thanks @afgane

nuwang commented:
PR 11 looks really nice - a much needed improvement over the current state. It will at present break some downstream components (e.g. The GVL microbial flavour that Simon authored relies on being able to specify multiple tool files), but I think that's ok - since we can wait for 2.0's imminent support for include+with_items. An ansible module would certainly be very nice - seems tantalisingly close :-)
